### DISCUSSION TITLE
How regular triangulation handles this special case?

### DISCUSSION THREAD

**Discussion by ningnawang** [CGAL/cgal#5763] [Q&A] [answered]
Hi CGALers,

I'm trying to use Regular_triangulation_2 and Regular_triangulation_3 to work on some interesting problems. However, I do not understand how regular triangulation (RT) in CGAL work for some special cases, would be really appreciated if someone'd like to share the answer. Here's the problem:

Given a set of weighted points S = (x, r^2), where x is the sphere/ball center and r is the radius. I want to construct the regular triangulation out of these spheres/balls.

I know that the **power diagram is dual to regular triangulation**. For general cases like below picture in 2D, circles A, B, C will form a RT face (red) because their power diagram (yellow) joins on a point p, so point p has equal power distance to A, B, or C.

![pic 1: general case](https://user-images.githubusercontent.com/8573845/120864039-95262b80-c559-11eb-8f01-8b33e96d425c.png)

However for some special cases: more than 3 circles whose power diagrams join on one point (more than 3 spheres join on one edge if 3D). For example the below picture, the power diagram (yellow) of circles 1,2,3,4,5,6 all join on point p. 

How would the regular triangulation in CGAL handle this case?

(I assume CGAL would split the red polygon to multiple triangles, but I do not know the rule).

![pic 2: special case](https://user-images.githubusercontent.com/8573845/120864072-a111ed80-c559-11eb-8feb-7e1ed47098c7.png)

Thank you in advance!!

**Comment by MaelRL** [comment, marked as answer]
In degenerate configurations where more than `k+1` (with `k` the dimension) vertices are on the same Delaunay ball, an arbitrary (but deterministic) triangulation is chosen.

In the code, this happens here: https://github.com/CGAL/cgal/blob/master/Triangulation_2/include/CGAL/Regular_triangulation_2.h#L869. The so-called "symbolic perturbation" virtually moves vertices a tiny bit as to get a non-degenerate configuration where the vertices of the polygon would no longer be all on the same Delaunay ball, which gives the triangulation of that polygon.
You can have also a look at the following publication: https://hal.inria.fr/inria-00560388/PDF/hal.pdf.

Note that your second figure seems to have mixed up the diagram and the triangulation.

**Reply by ningnawang** [reply]
Thank you for sharing!!!
